Add support for configuring ENI Connection Tracking timeouts

jukie opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.

Please review the Karpenter contribution docs at https://karpenter.sh/docs/contributing/ before submitting your pull request.
-->

~Fixes #8070~

Fixes #6826

**Description**
Allows for configuring ENI Connection Tracking timeouts

**How was this change tested?**
* Suite tests
* Manually in an EKS cluster
  * Updated EC2NodeClass CRD with the PR changes
  * Created test NodePool and EC2NodeClass with `connectionTracking` configured
  * created then a pod targeting the NodePool.
  * Validated ENI Connection Tracking timeouts were configured via AWS Console

**Does this change impact docs?**
- [x] Yes, PR includes docs updates <!-- docs must be added to /preview to be included in future version releases -->

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jukie committed:
Add support for configuring Connection Tracking settings

Signed-off-by: jukie <10012479+Jukie@users.noreply.github.com>

jukie committed:
Update docs and adjust descriptions

Signed-off-by: jukie <10012479+Jukie@users.noreply.github.com>

jukie committed:
rebase and feedback

Signed-off-by: jukie <10012479+Jukie@users.noreply.github.com>

jukie commented:
Hey @jonathan-innis could I have a review here please?

jukie commented:
@jonathan-innis could you please take a look at this or if you don't have time assign another reviewer?

jukie committed:
remove helper func

Signed-off-by: jukie <10012479+Jukie@users.noreply.github.com>

jukie committed:
cel validation

Signed-off-by: jukie <10012479+Jukie@users.noreply.github.com>

jukie commented on pkg/providers/launchtemplate/suite_test.go:
```diff
@@ -2457,6 +2459,41 @@ eviction-max-pod-grace-period = 10
 				Entry("AssociatePublicIPAddress is set as false and EFA is false", true, false, false),
 			)
 		})
+		Context("Connection Tracking settings", func() {
```
Updated, thanks

jukie commented on pkg/apis/v1/ec2nodeclass.go:
```diff
@@ -453,6 +460,27 @@ const (
+	// state.
+	// Value must be between 60 seconds and 432,000 seconds (5 days).
+	// +optional
+	TCPEstablishedTimeout *metav1.Duration `json:"tcpEstablishedTimeout,omitempty"`
```
I'd prefer to leave this as a duration field rather than strictly seconds. I've added CEL validation and it shouldn't create edge cases.

jukie commented on pkg/apis/v1/ec2nodeclass.go:
```diff
@@ -453,6 +460,27 @@ const (
+type ConnectionTracking struct {
+	// TCPEstablishedTimeout is the timeout for idle TCP connections in an established
+	// state.
+	// Value must be between 60 seconds and 432,000 seconds (5 days).
```
Sure, updated

jukie commented on pkg/apis/v1/ec2nodeclass.go:
```diff
@@ -138,6 +138,13 @@ type EC2NodeClassSpec struct {
+	// for AWS Elastic Network Interfaces. Idle connections left too long can
+	// exhaust the security group’s connection tracking table and lead to dropped packets.
+	// +optional
+	ConnectionTracking *ConnectionTracking `json:"connectionTracking,omitempty"`
```
Sure, added

jukie commented:
Sorry for the delayed updated but ready for re-review @engedaam!

jukie committed:
Merge branch 'main' into conntrack-settings

Commit:
Fix test

Signed-off-by: jukie <10012479+jukie@users.noreply.github.com>

jukie commented:
Updated again @engedaam if you can take a look after the holidays please

jukie committed:
Merge branch 'main' into conntrack-settings

jukie committed:
Merge branch 'main' into conntrack-settings

jukie committed:
Merge branch 'main' into conntrack-settings

jmdeal commented on website/content/en/preview/concepts/nodeclasses.md:
```diff
@@ -993,6 +1002,31 @@ spec:
+
+{{% alert title="Note" color="primary" %}}
+Changing any `connectionTracking` field will cause Karpenter to detect drift and replace existing nodes with new ones that have the updated configuration.
+{{% /alert %}}
```
I'm not sure this is worth including since we would need to call it out for basically every field. Notes add a lot of noise to the docs - honestly we overuse them already IMO. It would be worth consider a lower noise way to indicated how a given field interacts with drift though and use it consistently throughout the docs.

Zheaoli commented:
May I ask what is the status about this PR?

jukie commented:
I'll update it

jukie committed:
Merge branch 'main' into conntrack-settings

Signed-off-by: jukie <10012479+jukie@users.noreply.github.com>

jukie commented on pkg/providers/launchtemplate/launchtemplate.go:
```diff
@@ -288,6 +288,31 @@ func generateNetworkInterfaces(options *amifamily.LaunchTemplate, clusterIPFamil
 				AssociatePublicIpAddress: options.AssociatePublicIPAddress,
 				PrimaryIpv6:              lo.Ternary(clusterIPFamily == corev1.IPv6Protocol, lo.ToPtr(true), nil),
 				Ipv6AddressCount:         lo.Ternary(clusterIPFamily == corev1.IPv6Protocol, lo.ToPtr(int32(1)), nil),
+				ConnectionTrackingSpecification: func() *ec2types.ConnectionTrackingSpecificationRequest {
```
Karpenter has no control over secondary ENIs, so coverage there depends on what CNI the user runs.
- The primary ENI and EFA are controlled by the launch template, which Karpenter owns.
- Secondary ENIs are managed by the user's CNI.

With that in mind, I think it's reasonable to document the ownership scope and make users aware that secondary ENIs need to be configured elsewhere. For VPC-CNI specifically, there's in-flight work to add support with aws/amazon-vpc-cni-k8s#3618

jukie commented on test/suites/integration/connection_tracking_test.go:
```diff
@@ -0,0 +1,120 @@
+			NetworkInterfaceIds: []string{aws.ToString(primaryNI.NetworkInterfaceId)},
+		})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(niOutput.NetworkInterfaces).To(HaveLen(1))
```
Will add an EFA integration test. Secondary ENIs will be managed by the CNI so are out of scope

jukie committed:
refactor: refine conntrack settings based on review
- Extract connectionTrackingSpec() helper and call it from all three generateNetworkInterfaces paths (user-configured NetworkInterfaces, EFA loop, default single-ENI).
- Update field/type docstrings and the nodeclasses.md section to state scope explicitly: the spec applies to ENIs Karpenter provisions in the launch template (primary + EFA). Secondary ENIs created at runtime by the CNI are out of scope; point to aws/amazon-vpc-cni-k8s#3618 for the VPC-CNI companion work.
- Note that unset fields fall back to EC2 defaults.
- Remove the drift alert note.
- Drop the "partial configuration" integration test in favor of the existing unit coverage.
- Add an EFA integration test that iterates every ENI on the launched instance and asserts the spec is applied.

Signed-off-by: jukie <10012479+jukie@users.noreply.github.com>

jukie committed:
Merge branch 'main' into conntrack-settings

jukie committed:
cleanup

Signed-off-by: jukie <10012479+jukie@users.noreply.github.com>

jaydeokar commented on charts/karpenter-crd/templates/karpenter.k8s.aws_ec2nodeclasses.yaml:
```diff
@@ -295,6 +295,40 @@ spec:
+                        TCPEstablishedTimeout is the timeout for idle TCP connections in an established
+                        state.
+                        Value must be between 60 seconds and 432,000 seconds (5 days).
+                        If unset, EC2 applies its default of 432,000 seconds (5 days).
```
Not true for newer instances.. We should remove this statement
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/security-group-connection-tracking.html#connection-tracking-timeouts

jaydeokar commented:
@jukie Can you resolve the merge conflicts with this PR ?

sergiy-slobodyan commented:
Fixed by https://github.com/aws/karpenter-provider-aws/pull/9152
